smack-websocket/src/main/java/org/jivesoftware/smack/websocket/rce/WebsocketRemoteConnectionEndpoints.java

```diff
@@ -0,0 +1,100 @@
+
+        List<URI> rcUriList = null;
+        try {
+            // Look for remote connection endpoints by making use of http lookup method described inside XEP-0156.
+            rcUriList = HttpLookupMethod.lookup(connectionInternal.connection.getXMPPServiceDomain(),
+                            LinkRelation.WEBSOCKET);
+        } catch (IOException | XmlPullParserException | URISyntaxException e) {
+            lookupFailures.add(new RemoteConnectionEndpointLookupFailure.HttpLookupFailure(
+                            connectionInternal.connection.getXMPPServiceDomain(), e));
+            return new Result<>(discoveredRemoteConnectionEndpoints, lookupFailures);
+        }
+
+        if (rcUriList.isEmpty()) {
+            throw new IllegalStateException("No endpoints were found inside host-meta");
+        }
+
+        // Get the expected security mode for connection.
+        SecurityMode mode = connectionInternal.connection.getConfiguration().getSecurityMode();
+
+        if (mode.equals(SecurityMode.ifpossible)) {
+            // If security mode equals `if-possible`, give priority to secure endpoints over insecure endpoints.
+
+            // Seprate secure and unsecure endpoints.
+            List<WebsocketRemoteConnectionEndpoint> secureEndpointsForSecurityModeIfPossible = new ArrayList<>();
+            List<WebsocketRemoteConnectionEndpoint> insecureEndpointsForSecurityModeIfPossible = new ArrayList<>();
+            for (URI uri : rcUriList) {
+                WebsocketRemoteConnectionEndpoint endpoint = new WebsocketRemoteConnectionEndpoint(uri);
+                if (endpoint.isSecureEndpoint()) {
+                    secureEndpointsForSecurityModeIfPossible.add(endpoint);
+                } else {
+                    insecureEndpointsForSecurityModeIfPossible.add(endpoint);
+                }
+            }
+            discoveredRemoteConnectionEndpoints = secureEndpointsForSecurityModeIfPossible;
+            discoveredRemoteConnectionEndpoints.addAll(insecureEndpointsForSecurityModeIfPossible);
+        } else {
+            /**
+             * If, SecurityMode equals to required, accept wss endpoints (secure endpoints) only or,
+             * if SecurityMode equals to disabled, accept ws endpoints (unsecure endpoints) only.
+             */
+            for (URI uri : rcUriList) {
+                WebsocketRemoteConnectionEndpoint endpoint = new WebsocketRemoteConnectionEndpoint(uri);
+                if ((mode.equals(SecurityMode.disabled) && !endpoint.isSecureEndpoint())
+                                || (mode.equals(SecurityMode.required) && endpoint.isSecureEndpoint())) {
+                    discoveredRemoteConnectionEndpoints.add(endpoint);
+                }
+            }
+        }
+
+        return new Result<WebsocketRemoteConnectionEndpoint>(discoveredRemoteConnectionEndpoints, lookupFailures);
+    }
+
+    public static final class Result<WebsocketConnectionEndpoint> {
```

Comment:
There is no type `WebsocketConnectionEndpoint` in the codebase. The Generic annotation is useless. Simply remove it.